packages/MonorepoBuilder/packages/Split/tests/AbstractContainerAwareTestCase.php

```diff
@@ -23,9 +23,7 @@
      */
     public function __construct(?string $name = null, array $data = [], string $dataName = '')
     {
-        if (! self::$cachedContainer) {
-            self::$cachedContainer = (new ContainerFactory())->create();
-        }
+        self::$cachedContainer = (new ContainerFactory())->create();
```

Comment:
This kind off kills the caching, since the property is re-created everytime

Reply:
No, every time that we instantiate a new object, the `$cachedContainer` never exists

Reply:
On 1st run no, but this is cache.
On 2 run it exists.

Reply:
The correct fix is to add `|null` to the `$cachedContainer` property.

Reply:
Well, the value is never null after constructor.

Reply:
@TomasVotruba That's irrelevant since the property is static. It's null until the constructor is called. Yes, you can't really access it without calling the constructor because it's private and doesn't have a getter but it is still null.

Reply:
Non-static properties should not have `|null` because the only time you can access them while they have null value is in the constructor of the class. The annotations *of non-static properties* should reflect a state after calling __construct.

Static properties on the other hand are initialized when the php files is included. The only way to initialize them is dirrect assignment in the definition `public static $array = [];`. Otherwise they already exist in memory and have null value. `__construct` is not relevant here because you can access them without calling it.

Reply:
So calling a property in a constructor makes it have `|null` in annotation?

What if I call:

```php
public function __construct()
{
    if ($this->property === null) {
        // ...
    }
}
```

I think annotation should respect worflow as well, rather then pure logic.
**It's for programmers that read a code**, not for the compiler nor PHPStan.

When I read code and I see `|null` annotation, I know I have to check the nullable value everytime I use it. This is not the case.

Reply:
> So calling a property in a constructor makes it have |null in annotation?

No, it's the other way around: Not initializing a non-static property in constructor makes it have `|null` (unless there is a default value). Calling `if ($this->property === null) {` in `__construct` is useless for non static properties - you know that the value is null, there is no need to test it. This is however not the case for static properties as this use-case clearly shows.

> When I read code and I see |null annotation, I know I have to check the nullable value everytime I use it. This is not the case.

Congratulations - you just described the use-case we're commenting on - you are checking for null every time you use this property and therefore it should have `|null`.

Reply:
...everytime I use it in another method then constructor.

I'm still not convinced, due worse reability.

Reply:
The point is that there is no constructor for static properties.

Even with the logic you have in `__construct` if you accessed the property in a static method you would have to check for null. The fact that you're not doing that doesn't allow you to remove the `|null`.

Reply:
By the way in my opinion the `|null` actually improves readability in this case.

Reply:
```
class A {
    /** @var DateTimeImmutable */
    private static $now;

    public function __construct() {
        self::$now = new DateTimeImmutable();
    }

    public static function addDay(): void {
        // Oh cool, `self::$now` has type DateTimeImmutable so I can do this, right?
        self::$now = self::$now->modify('+1 day');
    }
}

A::addDay(); // Wait why is it failing??
```